embedded/tbtree/tbtree.go

```diff
@@ -1760,63 +1766,105 @@ func (t *TBtree) snapshotClosed(snapshot *Snapshot) error {
 	return nil
 }
 
-func (n *innerNode) insertAt(key []byte, value []byte, ts uint64) (nodes []node, depth int, err error) {
-	if !n.mutated() {
-		return n.copyOnInsertAt(key, value, ts)
+func (n *innerNode) insertAt(kvs []*KV, ts uint64) (nodes []node, depth int, err error) {
+	if n.mutated() {
+		return n.updateOnInsertAt(kvs, ts)
 	}
-	return n.updateOnInsertAt(key, value, ts)
-}
 
-func (n *innerNode) updateOnInsertAt(key []byte, value []byte, ts uint64) (nodes []node, depth int, err error) {
-	insertAt := n.indexOf(key)
+	newNode := &innerNode{
+		t:       n.t,
+		nodes:   make([]node, len(n.nodes)),
+		_ts:     n._ts,
+		_minOff: n._minOff,
+		mut:     true,
+	}
 
-	c := n.nodes[insertAt]
+	copy(newNode.nodes, n.nodes)
 
-	cs, depth, err := c.insertAt(key, value, ts)
-	if err != nil {
-		return nil, 0, err
+	return newNode.updateOnInsertAt(kvs, ts)
+}
+
+func (n *innerNode) updateOnInsertAt(kvs []*KV, ts uint64) (nodes []node, depth int, err error) {
+	// group kvs by child at which they will be inserted
+	kvsPerChild := make(map[int][]*KV)
+
+	for _, kv := range kvs {
+		childIndex := n.indexOf(kv.K)
+		kvsPerChild[childIndex] = append(kvsPerChild[childIndex], kv)
 	}
 
-	n._ts = ts
+	var wg sync.WaitGroup
+	wg.Add(len(kvsPerChild))
 
-	ns := make([]node, len(n.nodes)+len(cs)-1)
+	nodesPerChild := make(map[int][]node)
```

Comment:
it would be nice if you make this in a struct too, with lock/unlock methods which will make it easier to read/test